### PR TITLE
Change `find_node` to `find_nodes` and Add an `type` parameter.

### LINKED ISSUE #3661: Add a `Node.get_children_of_type()` method to get child nodes inheriting from a specific type
### Describe the project you are working on

N/A

### Describe the problem or limitation you are having in your project

I was thinking about how Interfaces are one of the lacking/missing OOP design patterns in Godot, but understood such a thing could be done via Composition over Inheritance with Child Nodes being the Interfaces that a Parent Node implements, and of course to use Interfaces one needs to know if something implements it, which led me to realize there is no helper function like `get_node()` for child nodes of specific types.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

```gdscript
var interfaces = get_children_of_type("Interface")
if interfaces.size() > 0:
    var interface = interfaces[0] as Interface
    interface.interface_func()
```

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

This but in C++:
```gdscript
var interfaces = []
for i in get_children():
    if i is Interface:
        interfaces.append(i)
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Yes, its simply a helper function such as `get_node()`.

### Is there a reason why this should be core and not an add-on in the asset library?

Less static functions in globals, and could see high use, but that's what this proposal is for.

### PR DESCRIPTION
**Pull request by Diddykonga**
Changed `find_node` to `find_nodes` which now returns an `TypedArray<Node>`, and Added an `type` parameter to match against specific node types, which supports inheritance. `find_nodes` can be used with just `mask` or just `type` or both.

`type`: Can be any `native`, `class_name`, or `resource_path` type, similarly to the `extends` keyword. (supports inheritance)

3.x PR: #56085
Closes: godotengine/godot-proposals#3661
Example Project: [FindNodesTest.zip](https://github.com/godotengine/godot/files/8063763/FindNodesTest.zip)

### CODE REVIEW

**Comment by me2beats** [comment]
Should it be named get_children_of_class() instead?

**Comment by nathanfranke** [comment]
> Should it be named get_children_of_class() instead?

`type` makes more sense in my opinion.

**Comment by akien-mga** [comment]
Something went wrong in the rebase and you added 353 unwanted commits.

**Comment by akien-mga** [comment]
That's worse now, you should now revert commits, you should remove them (with a `git rebase`). Reverting *adds* commits.

**Comment by diddykonga** [comment]
> 

Yeah its fixed my bad, I rebased to master, but couldn't compile due to some error in sw_renderer, so I atempted to rebase back  to my origin/master and it got goofed. I reset the branch back to the only commit.

**Comment by lawnjelly** [comment]
To be a lot more useful, imo this should include the option to search recursively through children's children etc. Me and @Calinou in fact previously discussed such a feature in terms of having the ability to right click in the editor in the scene tree on a branch, and select "find nodes of type".

See `find_node()` for an example of this.

So if this were the case, the name "children" wouldn't make so much sense versus e.g. "find_nodes_by_type". It would seem to make little sense to do a specific limited function that could only handle immediate children when long term we would probably want a recursive version available (no point in having two functions, or breaking the existing API, better to think ahead).

**Comment by diddykonga** [comment]
> To be a lot more useful, imo this should include the option to search recursively through children's children etc. Me and @Calinou in fact previously discussed such a feature in terms of having the ability to right click in the editor in the scene tree on a branch, and select "find nodes of type".
> 
> See `find_node()` for an example of this.
> 
> So if this were the case, the name "children" wouldn't make so much sense versus e.g. "find_nodes_by_type". It would seem to make little sense to do a specific limited function that could only handle immediate children when long term we would probably want a recursive version available (no point in having two functions, or breaking the existing API, better to think ahead).

Funny that you mention it, as you were writing this comment, I was including the option to do this recursively. xD

File: doc/classes/Node.xml
Comment (KoBeWi) [inline doc/classes/Node.xml:239]:
```suggestion
				[code]type[/code] could be a native type, [code]class_name[/code] type, or [code]resource_path[/code] type, similarly to the [code]extends[/code] keyword.
```

File: scene/main/node.cpp
Comment (KoBeWi) [inline scene/main/node.cpp:1247]: Empty type makes no sense for this method. I'd change it to `ERR_FAIL_COND_V`

File: scene/main/node.cpp
Comment (KoBeWi) [inline scene/main/node.cpp:1260]: This `continue` is not necessary.

**Comment by KoBeWi** [comment]
Looks good overall, unfortunately `recursive` option can't be tested due to #57008.

**Comment by diddykonga** [comment]
All suggested changes made, and rebased to current Master.

**Review by KoBeWi — APPROVED** [review]
Still needs a review from core team probably.

**Comment by lawnjelly** [comment]
> Still needs a review from core team probably.

Yes added PR meeting label as I think it needs broader review before going further.

On the naming at least I would be considering something like "find_nodes_of_type" rather than "get_children_of_type".

The difference between get and find seems already set by precedent in the `get_node()` and `find_node()` functions ("get" looking for a specific thing known in advance, and "find" implying a check on many nodes and potentially being more expensive .. this is often used as a convention for naming, with "get" used for accessors and "find" and "calculate" used for more expensive operations).

And children could be confusing if you are using the function recursively, which may end up being a common use case.

**Review by mhilbrunner — APPROVED** [review]

**Comment by akien-mga** [comment]
We discussed this in a PR review meeting, the added feature seems fine, but we were wondering if it wouldn't be worth refactoring this API to unify the existing `find_node` and the new functionality to restrict only to a specific type (i.e. make it `find_nodes` that can take an optional type parameter).

**Comment by Zireael07** [comment]
@akien-mga: Maybe as another PR? This is a feature that would be welcome sooner rather than later

**Comment by akien-mga** [comment]
> @akien-mga: Maybe as another PR? This is a feature that would be welcome sooner rather than later

Godot 4.0 is in alpha, and the further changes required mean breaking compatibility. There's no point merging a feature now in a pre-release branch to break compatibility right after.

**Comment by Zireael07** [comment]
Ah, good point.

**Comment by diddykonga** [comment]
Suggested changed from the PR Review Meeting have been implemented. I will wait for confirmation that the changes are correct/desired, before pushing the same changes to the 3.x PR.

File: editor/import/scene_import_settings.cpp
Comment (akien-mga) [inline editor/import/scene_import_settings.cpp:383]: `operator Node *()` was removed in #57205, but you can use `operator Object *()` which should work fine too.

File: doc/classes/Node.xml
Comment (akien-mga) [inline doc/classes/Node.xml:195]: The `<` and `>` break the XML: https://github.com/godotengine/godot/runs/5191971482?check_suite_focus=true
They would have to be escaped with `&lt;` and `&gt;`.

But here in the scripting (mostly GDScript-focused) documentation this is not a syntax that is part of the API (and `TypedArray` is not exposed as is), so this is probably best with a rewording (just "empty array" might be fine, the `Node[]` return type is already part of the signature - though here also not really something that can be used as is in the API, the GDScript type hint seems to be `Array[Node]`).

File: doc/classes/Node.xml
Comment (akien-mga) [inline doc/classes/Node.xml:192]:
```suggestion
				[b]Note:[/b] [code]mask[/code] does not match against the full path, just against individual node names. It is case-sensitive, with [code]"*"[/code] matching zero or more characters and [code]"?"[/code] matching any single character except [code]"."[/code]).
```

File: doc/classes/Node.xml
Comment (akien-mga) [inline doc/classes/Node.xml:193]:
```suggestion
				[code]type[/code] will check equality or inheritance. It is case-sensitive, [code]"Object"[/code] will match a node whose type is [code]"Node"[/code] but not the other way around.
```

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution :tada:

**Comment by codecat** [comment]
I'm seeing the recursive option not working as intended. For example, `find_nodes("*")` only returns the top nodes for me, not any child nodes. Similarly, searching for a class with either `find_nodes("*", "MyClass")` or `find_nodes("", "MyClass")` returns nothing, while there definitely are nodes of that type.

**Comment by bitbotzgames** [comment]
Why isn't there also a `find_parents` function?

**Comment by nathanfranke** [comment]
> Why isn't there also a `find_parents` function?

You can benefit from type hints with your own function, although I recognize the use case.

```gdscript
func get_map(n: Node = self) -> Map:
    if n == null:
        return null
    if n is Map:
        return n
    return get_map(n.get_parent())
```

Any other use cases that really shine?

**Comment by actforjason** [comment]
The parameter name should be `class` (from GetClass) rather than `type` (from GetType)

**Pull request merged**
